services/worker/src/worker/job_runners/split/opt_int_out_urls_scan_from_streaming.py

```diff
@@ -0,0 +1,430 @@
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "InfoError", cause, True)
+
+
+class TooManyColumnsError(OptinOutUrlsScanJobRunnerError):
+    """Raised when the dataset exceeded the max number of columns."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "TooManyColumnsError", cause, True)
+
+
+class PreviousStepStatusError(OptinOutUrlsScanJobRunnerError):
+    """Raised when the previous step gave an error. The job should not have been created."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "PreviousStepStatusError", cause, False)
+
+
+class PreviousStepFormatError(OptinOutUrlsScanJobRunnerError):
+    """Raised when the content of the previous step has not the expected format."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "PreviousStepFormatError", cause, False)
+
+
+class StreamingRowsError(OptinOutUrlsScanJobRunnerError):
+    """Raised when the rows could not be fetched in streaming mode."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "StreamingRowsError", cause, True)
+
+
+class NormalRowsError(OptinOutUrlsScanJobRunnerError):
+    """Raised when the rows could not be fetched in normal mode."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "NormalRowsError", cause, True)
+
+
+class MissingSpawningTokenError(OptinOutUrlsScanJobRunnerError):
+    """Raised when the spawning.ai token is not set."""
+
+    def __init__(self, message: str, cause: Optional[BaseException] = None):
+        super().__init__(message, HTTPStatus.INTERNAL_SERVER_ERROR, "MissingSpawningTokenError", cause, False)
+
+
+class OptinOutUrlsScanResponse(TypedDict):
+    urls_columns: List[str]
+    opt_in_urls: List[str]
+    opt_out_urls: List[str]
+    num_scanned_rows: int
+
+
+class OptInUrl(TypedDict):
+    url: str
+    row_idx: int
+    column_name: str
+
+
+class OptOutUrl(TypedDict):
+    url: str
+    row_idx: int
+    column_name: str
+
+
+@retry
+def get_rows(
+    dataset: str,
+    config: str,
+    split: str,
+    streaming: bool,
+    rows_max_number: int,
+    use_auth_token: Union[bool, str, None] = False,
+    column_names: Optional[List[str]] = None,
+) -> List[Row]:
+    download_config = DownloadConfig(delete_extracted=True)
+    ds = load_dataset(
+        dataset,
+        name=config,
+        split=split,
+        streaming=streaming,
+        use_auth_token=use_auth_token,
+        download_config=download_config,
+    )
```

Comment:
One optimization we could add at one point: if the builder is a Parquet builder, we can pass `columns=` as an additional parameter to only load the requested `column_names`.

Something like 
```python
builder = load_dataset_builder(
        dataset,
        name=config,
        use_auth_token=use_auth_token,
        download_config=download_config,
    )
builder_kwargs = {"columns": column_names} if isinstance(builder, Parquet) else {}
```
and adapt the rest to use `builder.as_streaming_dataset` or `.download_and_prepare` + `.as_datasaset()`